src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -0,0 +1,190 @@
+  _resize_increment(YoungGenerationSizeIncrement / 100.0),
+  _mmu_periodic_task(new ShenandoahMmuTask(this)),
+  _mmu_average(10, ShenandoahAdaptiveDecayFactor) {
+}
+
+ShenandoahMmuTracker::~ShenandoahMmuTracker() {
+  _mmu_periodic_task->disenroll();
+  delete _mmu_periodic_task;
+}
+
+void ShenandoahMmuTracker::record(ShenandoahGeneration* generation) {
+  // This is only called by the control thread.
+  double collector_time_s = gc_thread_time_seconds();
+  double elapsed_gc_time_s = collector_time_s - _initial_collector_time_s;
+  generation->add_collection_time(elapsed_gc_time_s);
+  _initial_collector_time_s = collector_time_s;
+}
+
+void ShenandoahMmuTracker::report() {
+  // This is only called by the periodic thread.
+  double process_time_s = process_time_seconds();
+  double elapsed_process_time_s = process_time_s - _initial_process_time_s;
+  _initial_process_time_s = process_time_s;
+  double verify_time_s = gc_thread_time_seconds();
+  double verify_elapsed = verify_time_s - _initial_verify_collector_time_s;
+  _initial_verify_collector_time_s = verify_time_s;
+  double verify_mmu = ((elapsed_process_time_s - verify_elapsed) / elapsed_process_time_s) * 100;
+  _mmu_average.add(verify_mmu);
+  log_info(gc)("Average MMU = %.3f", _mmu_average.davg());
+}
+
+bool ShenandoahMmuTracker::adjust_generation_sizes() {
+  shenandoah_assert_generational();
+  if (_mmu_average.davg() >= double(GCTimeRatio)) {
+    return false;
+  }
+
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  ShenandoahOldGeneration *old = heap->old_generation();
+  double old_time_s = old->reset_collection_time();
+  ShenandoahYoungGeneration *young = heap->young_generation();
+  double young_time_s = young->reset_collection_time();
+  ShenandoahGeneration *global = heap->global_generation();
+  double global_time_s = global->reset_collection_time();
+
+  log_info(gc)("Thread Usr+Sys YOUNG = %.3f, OLD = %.3f, GLOBAL = %.3f", young_time_s, old_time_s, global_time_s);
+
+  if (old_time_s > young_time_s) {
+    return transfer_capacity(young, old);
+  } else {
+    return transfer_capacity(old, young);
+  }
+}
+
+size_t percentage_of_heap(size_t bytes) {
+  size_t heap_capacity = ShenandoahHeap::heap()->max_capacity();
+  assert(bytes < heap_capacity, "Must be less than total capacity");
+  return size_t(100.0 * double(bytes) / double(heap_capacity));
+}
+
+bool ShenandoahMmuTracker::transfer_capacity(ShenandoahGeneration* from, ShenandoahGeneration* to) {
+  shenandoah_assert_heaplocked_or_safepoint();
+
+  size_t available_regions = from->free_unaffiliated_regions();
+  if (available_regions <= 0) {
+    log_info(gc)("%s has no regions available for transfer to %s", from->name(), to->name());
+    return false;
+  }
+
+  size_t regions_to_transfer = MAX2(1UL, size_t(double(available_regions) * _resize_increment));
+  size_t bytes_to_transfer = regions_to_transfer * ShenandoahHeapRegion::region_size_bytes();
+  if (from->generation_mode() == YOUNG) {
```

Comment:
Yes, this method is a bit long. The similarity breaks a bit because it needs to enforce the min or max constraint depending on the direction of the transfer. I'll split it up in a subsequent PR.